File: subtracker2015.cpp
#include <iostream>

#include "control.hpp"
#include "framereader.hpp"
#include "context.hpp"

using namespace cv;
using namespace std;

static void key_pressed(SubtrackerContext &ctx, char c) {

  control_panel_t &panel = ctx.panel;

  switch(c) {

  case 'r':
    set_log_level(panel, "capture", VERBOSE);
    break;

  case 'p':
    toggle(panel, "control panel", TRACKBAR);
    break;

  case 'f':
    toggle(panel, "frame", SHOW);
    break;

  case 'c':
    toggle(panel, "cycle", TIME);
    break;

	case 's':
    set_log_level(panel, "table detect", VERBOSE);
    toggle(panel, "table detect", TRACKBAR, true);
    toggle(panel, "table detect", SHOW, true);
    break;

  case 'b':
    set_log_level(panel, "ball tracking", VERBOSE);
    toggle(panel, "ball tracking", TRACKBAR, true);
    toggle(panel, "ball tracking", SHOW, true);
    break;

  case 'm':
    toggle(panel, "foosmen tracking", TRACKBAR);
    toggle(panel, "foosmen tracking", SHOW);
    break;

  case 'l':
    toggle(panel, "foosmen metrics", TRACKBAR);
    toggle(panel, "reference metrics", TRACKBAR);
    break;

  case 'n':
    toggle(panel, "foosmen colors", TRACKBAR);
    break;

  case ' ':
    set_log_level(panel, "table detect", WARNING);
    set_log_level(panel, "ball tracking", WARNING);
    set_log_level(panel, "capture", WARNING);

    toggle(panel, "table detect", TRACKBAR, false);
    toggle(panel, "ball tracking", TRACKBAR, false);
    toggle(panel, "ball tracking", SHOW, false);

    toggle(panel, "foosmen tracking", TRACKBAR, false);
    toggle(panel, "foosmen tracking", SHOW, false);

    toggle(panel, "foosmen metrics", TRACKBAR, false);
    toggle(panel, "reference metrics", TRACKBAR, false);

    toggle(panel, "foosmen colors", TRACKBAR, false);

    toggle(panel, "frame", SHOW, false);
    toggle(panel, "cycle", TIME, false);
    toggle(panel, "control panel", TRACKBAR, false);
    break;
  }

}

int update_gui_skip = 5;

static void feed_frames(FrameReader &frame_reader, SubtrackerContext &ctx) {

  control_panel_t &panel = ctx.panel;

  trackbar(panel, "control panel", "update GUI skip", update_gui_skip, {0, 20, 1});

  for (int frame_num = 0; ; frame_num++) {

    dump_time(panel, "cycle", "feed new frame");

    // Every now and then, do GUI computation
    panel.update_display = (frame_num % (update_gui_skip + 1) == 0);
    if (panel.update_display) {

			namedWindow("control panel", WINDOW_NORMAL);

      // Check of key pressions (FIXME: this process just one key
      // pression every cycle)
			int c = waitKey(1);
      if (c > 0) c &= 0xff;
      key_pressed(ctx, c);

      dump_time(panel, "cycle", "wait key");

    }

    // Read a new frame and perhaps terminate program
    auto frame_info = frame_reader.get();
    if (!frame_info.valid) break;

    dump_time(panel, "cycle", "got a frame");

    // Feed the frame to the subtracker
    ctx.feed(frame_info.data, frame_info.timestamp, frame_info.playback_time);

    dump_time(panel, "cycle", "finished frame feed");

  }

}

int main(int argc, char* argv[]) {

  // Parse arguments
  string videoName, referenceImageName, referenceImageMaskName;
  if (argc == 3 || argc == 4) {
    videoName = argv[1];
    referenceImageName = argv[2];

    if(argc == 4) {
      referenceImageMaskName = argv[3];
    }
  } else {
    cerr << "Usage: " << argv[0] << " <video> <reference subotto> [<reference subotto mask>]" << endl;
    cerr << "<video> can be: " << endl;
    cerr << "\tn (a single digit number) - live capture from video device n" << endl;
    cerr << "\tfilename+ (with a trailing plus) - simulate live capture from video file" << endl;
    cerr << "\tfilename (without a trailing plus) - batch analysis of video file" << endl;
    cerr << "<reference subotto> is the reference image used to look for the table" << endl;
    cerr << "<reference sumotto mask> is an optional B/W image, of the same size," << endl;
    cerr << "\twhere black spots indicate areas of the reference image to hide" << endl;
    cerr << "\twhen looking for the table (such as moving parts and spurious features)" << endl;
    return 1;
  }

  control_panel_t panel;

  // Read reference image and mask
  Mat ref_frame = imread(referenceImageName);
  Mat ref_mask;
  if(!referenceImageMaskName.empty()) {
    ref_mask = imread(referenceImageMaskName, CV_LOAD_IMAGE_GRAYSCALE);
  }

  SubtrackerContext ctx(ref_frame, ref_mask, panel);

  // Initialize panel (GUI)
  init_control_panel(panel);
  set_log_level(panel, "gio", VERBOSE);

  // Open frame reader
  FrameReader *f;
  if(videoName.size() == 1) {
    f = new FrameReader(videoName[0] - '0', panel);
  } else {
    bool simulate_live = false;

    if(videoName.back() == '+') {
      videoName = videoName.substr(0, videoName.size()-1);
      simulate_live = true;
    }

    f = new FrameReader(videoName.c_str(), panel, simulate_live);
  }

  feed_frames(*f, ctx);

  // Graciously wait for the reader thread to stop
  delete f;

  return 0;

}

File: subtracker2015.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-

import sys
import cv2
import logging

from control import ControlPanel
from context import SubtrackerContext
from framereader import CameraFrameReader, FileFrameReader

logger = logging.getLogger("main program")


# Mock implementation
class MockControlPanel():

    def on_key_pressed(self, c):
        pass

    def on_new_analysis(self, frame_analysis):
        pass


def feed_frames(frame_reader, ctx, panel):
    frame_num = 0
    update_gui_skip = 1
    step_frame = False
    step_on_frame_produced = False
    while True:
        # Decide whether we have to update GUI; so far it always
        # happen!
        update_display = (frame_num % (update_gui_skip + 1) == 0) or step_frame
        if update_display:
            while True:
                #logger.debug("Before waitKey()")
                c = cv2.waitKey(1)
                if c < 0:
                    if not step_frame:
                        break
                    else:
                        continue
                c &= 0xff
                c = chr(c)
                #logger.debug("After waitKey() = %r", c)
                if step_frame and c == '.':
                    break
                if c == 'q':
                    raise KeyboardInterrupt
                panel.on_key_pressed(c)

        # Get a frame and feed it to the context
        frame_info = frame_reader.get()
        if not frame_info.valid:
            break
        ctx.feed(frame_info.frame, frame_info.timestamp, frame_info.playback_time)

        # Retrieve as many processed frames as possible
        while True:
            frame_analysis = ctx.get_processed_frame()
            if frame_analysis is None:
                break
            panel.on_new_analysis(frame_analysis)
            sys.stdout.write(frame_analysis.get_csv_line())
            if step_on_frame_produced:
                step_frame = True

        # Prepare for next round
        frame_num += 1

def main():
    logging.basicConfig(level=logging.INFO,
                        format='%(asctime)s.%(msecs).03d (%(name)s) %(message)s',
                        datefmt='%d/%m/%Y %H:%M:%S')

    # Read arguments
    if len(sys.argv) in [1,2]:
        video_filename = sys.argv[1]
    else:
        print >> sys.stderr, "Usage: %s <video>" % (sys.argv[0])
        print >> sys.stderr, "<video> can be: "
        print >> sys.stderr, "\tn (a single digit number) - live capture from video device n"
        print >> sys.stderr, "\tfilename+ (with a trailing plus) - simulate live capture from video file"
        print >> sys.stderr, "\tfilename (without a trailing plus) - batch analysis of video file"
        sys.exit(1)

    # Create panel and context
    panel = ControlPanel()
    ctx = SubtrackerContext(panel)

    # Create frame reader
    if len(video_filename) == 1:
        frame_reader = CameraFrameReader(ord(video_filename) - ord('0'))
    else:
        simulate_live = False
        if video_filename[-1] == '+':
            video_filename = video_filename[:-1]
            simulate_live = True
        frame_reader = FileFrameReader(video_filename, simulate_live)
    frame_reader.start()

    # Do the actual work
    try:
        feed_frames(frame_reader, ctx, panel)
    except KeyboardInterrupt:
        logger.info("Exit requested")
    except:
        logger.info("Exit because of exception", exc_info=True)

    # When finishing, wait for the reader thread to terminate
    # gracefully and consume left frames
    frame_reader.stop()
    while True:
        if frame_reader.get(block=False) is None:
            break
    frame_reader.join()

    logger.info("Exiting")

if __name__ == '__main__':
    main()
